src/electron/office/document_client.cc

```diff
@@ -330,4 +358,279 @@ void DocumentClient::Emit(const std::string& event_name,
+
+    in_sizes[i] = buffer->ByteLength();
+    mime_c_str.push_back(mime_type.c_str());
+    streams[i] = static_cast<char*>(buffer->GetBackingStore()->Data());
+  }
+
+  // add the nullptr terminator to the list of null-terminated strings
+  mime_c_str.push_back(nullptr);
+
+  return document_->setClipboard(entries, mime_c_str.data(), in_sizes, streams);
+}
+
+bool DocumentClient::Paste(const std::string& mime_type,
+                           const std::string& data,
+                           gin::Arguments* args) {
+  return document_->paste(mime_type.c_str(), data.c_str(), data.size());
+}
+
+void DocumentClient::SetGraphicSelection(int n_type, int n_x, int n_y) {
+  document_->setGraphicSelection(n_type, n_x, n_y);
+}
+
+void DocumentClient::ResetSelection() {
+  document_->resetSelection();
+}
+
+v8::Local<v8::Value> DocumentClient::GetCommandValues(
+    const std::string& command,
+    gin::Arguments* args) {
+  char* result = document_->getCommandValues(command.c_str());
+
+  v8::Isolate* isolate = args->isolate();
+
+  v8::MaybeLocal<v8::String> maybe_res_json_str =
+      v8::String::NewFromUtf8(isolate, result);
+
+  if (maybe_res_json_str.IsEmpty()) {
+    return v8::Undefined(isolate);
+  }
+
+  v8::MaybeLocal<v8::Value> res = v8::JSON::Parse(
+      args->GetHolderCreationContext(), maybe_res_json_str.ToLocalChecked());
+
+  if (res.IsEmpty()) {
+    return v8::Undefined(isolate);
+  }
+
+  return res.ToLocalChecked();
+}
+
+void DocumentClient::SetOutlineState(bool column,
+                                     int level,
+                                     int index,
+                                     bool hidden) {
+  document_->setOutlineState(column, level, index, hidden);
+}
+
+void DocumentClient::SetViewLanguage(int id, const std::string& language) {
+  document_->setViewLanguage(id, language.c_str());
+}
+
+void DocumentClient::SelectPart(int part, int select) {
+  document_->selectPart(part, select);
+}
+
+void DocumentClient::MoveSelectedParts(int position, bool duplicate) {
+  document_->moveSelectedParts(position, duplicate);
+}
+
+void DocumentClient::RemoveTextContext(unsigned window_id,
+                                       int before,
+                                       int after) {
+  document_->removeTextContext(window_id, before, after);
+}
+
+void DocumentClient::CompleteFunction(const std::string& function_name) {
+  document_->completeFunction(function_name.c_str());
+}
+
+void DocumentClient::SendFormFieldEvent(const std::string& arguments) {
+  document_->sendFormFieldEvent(arguments.c_str());
+}
+
+void DocumentClient::SendContentControlEvent(
+    const v8::Local<v8::Object>& arguments,
+    gin::Arguments* args) {
+  v8::Isolate* isolate = args->isolate();
+
+  // Object -> JSON string -> string
+  v8::MaybeLocal<v8::String> str_object =
+      v8::JSON::Stringify(args->GetHolderCreationContext(), arguments);
+
+  if (str_object.IsEmpty()) {
+  }
```

Comment:
This would be a good example of a failure where the value returned should be false, rather than throwing an error.

Reply:
This method returns void though so I think we should just return here. I must've had that there to throw an error but I just returning out is probably best